File: PhoneValidator.java
// PhoneValidator.java
import java.io.*;
import java.util.*;
import java.util.regex.*;

public class PhoneValidator {
    private static class CountryInfo {
        String name;
        int minLen, maxLen;
        Pattern pattern;
        CountryInfo(String n, int min, int max, String regex) {
            name = n; minLen = min; maxLen = max; pattern = Pattern.compile(regex);
        }
    }

    private static final Map<String, CountryInfo> COUNTRY_DATA = new HashMap<>();
    static {
        COUNTRY_DATA.put("1", new CountryInfo("United States/Canada", 10, 10, "^[2-9]\\d{9}$"));
        COUNTRY_DATA.put("44", new CountryInfo("United Kingdom", 10, 10, "^[1-9]\\d{9}$"));
        COUNTRY_DATA.put("49", new CountryInfo("Germany", 10, 11, "^[1-9]\\d{9,10}$"));
        COUNTRY_DATA.put("33", new CountryInfo("France", 9, 10, "^[1-9]\\d{8,9}$"));
        COUNTRY_DATA.put("7", new CountryInfo("Russia/Kazakhstan", 10, 10, "^[78]\\d{9}$"));
        COUNTRY_DATA.put("380", new CountryInfo("Ukraine", 9, 9, "^[0-9]{9}$"));
        COUNTRY_DATA.put("61", new CountryInfo("Australia", 9, 10, "^[0-9]{9,10}$"));
        COUNTRY_DATA.put("91", new CountryInfo("India", 10, 10, "^[6-9]\\d{9}$"));
        COUNTRY_DATA.put("55", new CountryInfo("Brazil", 10, 11, "^[1-9]\\d{9,10}$"));
        COUNTRY_DATA.put("86", new CountryInfo("China", 11, 11, "^1\\d{10}$"));
        COUNTRY_DATA.put("81", new CountryInfo("Japan", 10, 10, "^[0-9]{10}$"));
        COUNTRY_DATA.put("82", new CountryInfo("South Korea", 10, 11, "^[0-9]{10,11}$"));
    }

    private int total, valid, invalid;

    public PhoneValidator() {
        total = valid = invalid = 0;
    }

    public Result validateNumber(String raw) {
        total++;
        raw = raw.trim();
        if (raw.isEmpty()) { invalid++; return new Result(false, null, null, "Empty input"); }
        String cleaned = raw.replaceAll("[^\\d+]", "");
        if (cleaned.isEmpty()) { invalid++; return new Result(false, null, null, "No digits found"); }
        String countryCode = null;
        String restDigits = null;
        if (cleaned.startsWith("+")) {
            for (int i = 1; i <= 3 && i < cleaned.length(); i++) {
                String cc = cleaned.substring(1, i+1);
                if (COUNTRY_DATA.containsKey(cc)) {
                    countryCode = cc;
                    restDigits = cleaned.substring(i+1);
                    break;
                }
            }
            if (countryCode == null) { invalid++; return new Result(false, null, null, "Unknown country code"); }
        } else {
            String digits = raw.replaceAll("\\D", "");
            if (digits.length() == 10 && digits.matches("^[2-9]\\d{9}$")) {
                String normalized = "+1" + digits;
                valid++;
                return new Result(true, normalized, "United States/Canada", null);
            }
            invalid++;
            return new Result(false, null, null, "Could not determine country code; use + prefix");
        }

        CountryInfo info = COUNTRY_DATA.get(countryCode);
        if (info == null) { invalid++; return new Result(false, null, null, "Unknown country code"); }
        if (restDigits == null || restDigits.isEmpty()) { invalid++; return new Result(false, null, info.name, "Missing subscriber number"); }
        if (restDigits.length() < info.minLen || restDigits.length() > info.maxLen) {
            invalid++;
            return new Result(false, null, info.name, "Invalid length: " + restDigits.length() + " (expected " + info.minLen + "-" + info.maxLen + ")");
        }
        if (!info.pattern.matcher(restDigits).matches()) {
            invalid++;
            return new Result(false, null, info.name, "Invalid digit pattern");
        }
        String normalized = "+" + countryCode + restDigits;
        valid++;
        return new Result(true, normalized, info.name, null);
    }

    public List<Result> batchValidate(String[] numbers) {
        List<Result> results = new ArrayList<>();
        for (String n : numbers) {
            String raw = n.trim();
            if (raw.isEmpty()) continue;
            Result r = validateNumber(raw);
            r.original = raw;
            results.add(r);
        }
        return results;
    }

    public void showStats() {
        System.out.printf("\nStatistics: Total: %d, Valid: %d, Invalid: %d\n", total, valid, invalid);
    }

    static class Result {
        boolean valid;
        String normalized;
        String country;
        String error;
        String original;
        Result(boolean v, String norm, String c, String e) { valid = v; normalized = norm; country = c; error = e; }
    }

    public static void main(String[] args) throws IOException {
        PhoneValidator validator = new PhoneValidator();
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        System.out.println("=== Phone Number Validator ===");
        while (true) {
            System.out.println("\n1. Validate single number");
            System.out.println("2. Validate from file");
            System.out.println("3. Show statistics");
            System.out.println("4. Exit");
            System.out.print("Choose: ");
            String choice = reader.readLine().trim();
            switch (choice) {
                case "1":
                    System.out.print("Enter phone number: ");
                    String num = reader.readLine().trim();
                    Result res = validator.validateNumber(num);
                    System.out.println("Valid: " + res.valid);
                    if (res.valid) {
                        System.out.println("Normalized: " + res.normalized);
                        System.out.println("Country: " + res.country);
                    } else {
                        System.out.println("Error: " + res.error);
                    }
                    break;
                case "2":
                    System.out.print("Enter file path: ");
                    String fname = reader.readLine().trim();
                    File file = new File(fname);
                    if (!file.exists()) {
                        System.out.println("File not found.");
                        break;
                    }
                    List<String> lines = new ArrayList<>();
                    try (BufferedReader br = new BufferedReader(new FileReader(file))) {
                        String line;
                        while ((line = br.readLine()) != null) {
                            lines.add(line);
                        }
                    }
                    String[] arr = lines.toArray(new String[0]);
                    List<Result> results = validator.batchValidate(arr);
                    System.out.println("\nBatch results:");
                    for (Result r : results) {
                        String status = r.valid ? "✓" : "✗";
                        System.out.printf("%s %s: %s\n", status, r.original, r.error != null ? r.error : "OK");
                        if (r.valid) {
                            System.out.printf("   Normalized: %s, Country: %s\n", r.normalized, r.country);
                        }
                    }
                    break;
                case "3":
                    validator.showStats();
                    break;
                case "4":
                    System.out.println("Goodbye!");
                    return;
                default:
                    System.out.println("Invalid choice.");
            }
        }
    }
}

File: PhoneValidator.cs
// PhoneValidator.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

class CountryInfo {
    public string Name { get; set; }
    public int MinLen { get; set; }
    public int MaxLen { get; set; }
    public Regex Regex { get; set; }
}

class PhoneValidator {
    private static readonly Dictionary<string, CountryInfo> CountryData = new Dictionary<string, CountryInfo>
    {
        {"1", new CountryInfo{Name="United States/Canada", MinLen=10, MaxLen=10, Regex=new Regex(@"^[2-9]\d{9}$")}},
        {"44", new CountryInfo{Name="United Kingdom", MinLen=10, MaxLen=10, Regex=new Regex(@"^[1-9]\d{9}$")}},
        {"49", new CountryInfo{Name="Germany", MinLen=10, MaxLen=11, Regex=new Regex(@"^[1-9]\d{9,10}$")}},
        {"33", new CountryInfo{Name="France", MinLen=9, MaxLen=10, Regex=new Regex(@"^[1-9]\d{8,9}$")}},
        {"7", new CountryInfo{Name="Russia/Kazakhstan", MinLen=10, MaxLen=10, Regex=new Regex(@"^[78]\d{9}$")}},
        {"380", new CountryInfo{Name="Ukraine", MinLen=9, MaxLen=9, Regex=new Regex(@"^[0-9]{9}$")}},
        {"61", new CountryInfo{Name="Australia", MinLen=9, MaxLen=10, Regex=new Regex(@"^[0-9]{9,10}$")}},
        {"91", new CountryInfo{Name="India", MinLen=10, MaxLen=10, Regex=new Regex(@"^[6-9]\d{9}$")}},
        {"55", new CountryInfo{Name="Brazil", MinLen=10, MaxLen=11, Regex=new Regex(@"^[1-9]\d{9,10}$")}},
        {"86", new CountryInfo{Name="China", MinLen=11, MaxLen=11, Regex=new Regex(@"^1\d{10}$")}},
        {"81", new CountryInfo{Name="Japan", MinLen=10, MaxLen=10, Regex=new Regex(@"^[0-9]{10}$")}},
        {"82", new CountryInfo{Name="South Korea", MinLen=10, MaxLen=11, Regex=new Regex(@"^[0-9]{10,11}$")}}
    };

    public (int Total, int Valid, int Invalid) Stats { get; private set; }

    public PhoneValidator() {
        Stats = (0, 0, 0);
    }

    public (bool valid, string normalized, string country, string error) ValidateNumber(string raw) {
        Stats.Total++;
        raw = raw.Trim();
        if (string.IsNullOrEmpty(raw)) {
            Stats.Invalid++;
            return (false, null, null, "Empty input");
        }
        string cleaned = Regex.Replace(raw, @"[^\d+]", "");
        if (string.IsNullOrEmpty(cleaned)) {
            Stats.Invalid++;
            return (false, null, null, "No digits found");
        }
        string countryCode = null;
        string restDigits = null;
        if (cleaned.StartsWith("+")) {
            for (int i = 1; i <= 3 && i < cleaned.Length; i++) {
                string cc = cleaned.Substring(1, i);
                if (CountryData.ContainsKey(cc)) {
                    countryCode = cc;
                    restDigits = cleaned.Substring(i + 1);
                    break;
                }
            }
            if (countryCode == null) {
                Stats.Invalid++;
                return (false, null, null, "Unknown country code");
            }
        } else {
            // No country code: try US
            string digits = Regex.Replace(raw, @"\D", "");
            if (digits.Length == 10 && Regex.IsMatch(digits, @"^[2-9]\d{9}$")) {
                string normalized = "+1" + digits;
                Stats.Valid++;
                return (true, normalized, "United States/Canada", null);
            }
            Stats.Invalid++;
            return (false, null, null, "Could not determine country code; use + prefix");
        }

        if (!CountryData.ContainsKey(countryCode)) {
            Stats.Invalid++;
            return (false, null, null, "Unknown country code");
        }
        var info = CountryData[countryCode];
        if (string.IsNullOrEmpty(restDigits)) {
            Stats.Invalid++;
            return (false, null, info.Name, "Missing subscriber number");
        }
        if (restDigits.Length < info.MinLen || restDigits.Length > info.MaxLen) {
            Stats.Invalid++;
            return (false, null, info.Name, $"Invalid length: {restDigits.Length} (expected {info.MinLen}-{info.MaxLen})");
        }
        if (!info.Regex.IsMatch(restDigits)) {
            Stats.Invalid++;
            return (false, null, info.Name, "Invalid digit pattern");
        }
        string normalized = "+" + countryCode + restDigits;
        Stats.Valid++;
        return (true, normalized, info.Name, null);
    }

    public List<(string original, bool valid, string normalized, string country, string error)> BatchValidate(string[] numbers) {
        var results = new List<(string, bool, string, string, string)>();
        foreach (var n in numbers) {
            string raw = n.Trim();
            if (string.IsNullOrEmpty(raw)) continue;
            var res = ValidateNumber(raw);
            results.Add((raw, res.valid, res.normalized, res.country, res.error));
        }
        return results;
    }

    public void ShowStats() {
        Console.WriteLine($"\nStatistics: Total: {Stats.Total}, Valid: {Stats.Valid}, Invalid: {Stats.Invalid}");
    }

    static void Main() {
        var validator = new PhoneValidator();
        Console.WriteLine("=== Phone Number Validator ===");
        while (true) {
            Console.WriteLine("\n1. Validate single number");
            Console.WriteLine("2. Validate from file");
            Console.WriteLine("3. Show statistics");
            Console.WriteLine("4. Exit");
            Console.Write("Choose: ");
            string choice = Console.ReadLine()?.Trim() ?? "";
            switch (choice) {
                case "1":
                    Console.Write("Enter phone number: ");
                    string num = Console.ReadLine()?.Trim() ?? "";
                    var result = validator.ValidateNumber(num);
                    Console.WriteLine($"Valid: {result.valid}");
                    if (result.valid) {
                        Console.WriteLine($"Normalized: {result.normalized}");
                        Console.WriteLine($"Country: {result.country}");
                    } else {
                        Console.WriteLine($"Error: {result.error}");
                    }
                    break;
                case "2":
                    Console.Write("Enter file path: ");
                    string fname = Console.ReadLine()?.Trim() ?? "";
                    if (!File.Exists(fname)) {
                        Console.WriteLine("File not found.");
                        break;
                    }
                    string[] lines = File.ReadAllLines(fname);
                    var results = validator.BatchValidate(lines);
                    Console.WriteLine("\nBatch results:");
                    foreach (var r in results) {
                        string status = r.valid ? "✓" : "✗";
                        Console.WriteLine($"{status} {r.original}: {r.error ?? "OK"}");
                        if (r.valid) {
                            Console.WriteLine($"   Normalized: {r.normalized}, Country: {r.country}");
                        }
                    }
                    break;
                case "3":
                    validator.ShowStats();
                    break;
                case "4":
                    Console.WriteLine("Goodbye!");
                    return;
                default:
                    Console.WriteLine("Invalid choice.");
                    break;
            }
        }
    }
}
